addon-test-support/ember-mocha/setup-application-test.js

```diff
@@ -0,0 +1,21 @@
+import {
+  beforeEach,
+  afterEach
+} from 'mocha';
+import {
+  setupApplicationContext,
+  teardownApplicationContext
+} from '@ember/test-helpers';
+import setupUnitTest from './setup-unit-test';
+
+export default function setupApplicationTest(options) {
+  setupUnitTest(options);
+
+  beforeEach(function() {
+    return setupApplicationContext(this);
+  });
+
+  afterEach(function() {
+    return teardownApplicationContext(this);
+  });
```

Comment:
wouldn't this have to be written before the `setupUnitTest()` call due to the FIFO ordering?

Reply:
Hm, maybe yes, good catch! Right now it does not make a difference I think, because the [implementation of `teardownApplicationContext `](https://github.com/emberjs/ember-test-helpers/blob/master/addon-test-support/%40ember/test-helpers/teardown-application-context.js) does not depend on it. But we should probably not rely on this...

I suppose we won't have a solution that "just works" for the general ordering problem? (other than to document it properly)

Reply:
> I suppose we won't have a solution that "just works" for the general ordering problem?

I do have an idea...

```js
let hooks = setupTest();

hooks.beforeEach(function() { ... });
hooks.afterEach(function() { ... });
```

`setupTest()` maintains its own FIFO/LIFO queues and exposes them by returning an object with `before/afterEach` functions. The additional advantage is that the API becomes even closer to QUnit.

The disadvantage is that for Mocha users calling `hooks.afterEach` is unusual, but since this feature is only needed as an escape valve and probably shouldn't be needed by most endusers this might be fine 🤔

Reply:
As the user has to be made aware of the problem anyway, and be told to use this (as you also mention) rather unusual way, I wonder if it is not equally appropriate and less "magic" to just tell them to move their `afterEach` calls before the `setupTest()` call?

Reply:
> move their afterEach calls before the setupTest() call?

that is not an option if you want to provide e.g. a `setupPretender(hooks)` helper that needs to setup both hooks